[63475] Preserve unset dates when moving work packages in calendar

cbliard opened the pull request:
# Ticket

https://community.openproject.org/wp/63475

# What are you trying to accomplish?

Preserve unset dates when moving a work package with only one date set on the calendar. Moving it should not set both dates without explicit user intent, especially now that we have a single-date mode for work packages.

## Screenshots

None provided

# What approach did you choose and why?

If it's resized, both start and due dates are set.

If it's dragged and dropped and it's a milestone, it's the same as resizing: both dates are set.

If it's dragged and dropped and it's not a milestone, if it has a duration, set duration and start date to deal with non-working days, and if it has no duration, set start date if it was already set, or set the due date.

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

cbliard committed:
[63475] Preserve unset dates when moving work packages in calendar

https://community.openproject.org/wp/63475

If it's resized, both start and due dates are set.

If it's dragged and dropped and it's a milestone, it's the same as
resizing: both dates are set.

If it's dragged and dropped and it's not a milestone, if it has a
duration, set duration and start date to deal with non-working days, and
if it has no duration, set start date if it was already set, or set the
due date.

cbliard committed:
[63475] Set dates when dropping work packages in team planner

When drag&dropping a work package in the team planner from the left
panel, the dates should be set, especially if they were not set
initially. This differs from the behavior of the calendar module where
drag&dropping from one date to another should keep any unset dates.

Also, only start date and duration are set to be able to deal with
non-working days.

That use case was hopefully caught by a failing feature spec:
modules/team_planner/spec/features/team_planner_add_existing_work_packages_spec.rb:135.
I added another spec to ensure that non-working days are handled correctly.

klaustopher commented on modules/calendar/spec/features/calendar_user_interaction_spec.rb:
```diff
@@ -34,17 +34,16 @@
-  end
-
-  let!(:work_package) do
+  shared_let(:monday) { Time.zone.today.beginning_of_week }
```
This is only true, when `Date.beginning_of_week` is set to `:monday`. This might not be true on all systems, to ensure that we always have the monday, let's use:

```suggestion
  shared_let(:monday) { Time.zone.today.beginning_of_week(:monday) }
```

cbliard replied:
Sweet! I didn't know `beginning_of_week` accepted parameters. (Well, looking at [some of my commits](https://github.com/opf/openproject/commit/3b20568377cd097776923c90f99605848ef8e580#diff-b51c5cf8597049c3bdf710622f5fbcabaa166b99ce40c906e552e7f81eaa34bcR37), it looks like I knew it, and forgot it)
Very good suggestion, I'll add it.

cbliard committed:
More robust way of getting monday of current week

`Date.current.beginning_of_week` returns a Monday only if
`Date.beginning_of_week` is set to `:monday`, which may not be true on
all systems. We need to be explicit about it.

Co-authored-by: Klaus Zanders <k.zanders@openproject.com>
